File: demo/pyodide/check.mjs
// Runs peer_check.py under Pyodide in node, installing ontodag from PyPI
// through micropip exactly as a browser would. Prints the one JSON line.
//   npm install   (once; pyodide is the only dependency)
//   node check.mjs [VERSION]
import { loadPyodide } from "pyodide";
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const here = dirname(fileURLToPath(import.meta.url));
const spec = process.argv[2] ? `ontodag==${process.argv[2]}` : "ontodag";
const t0 = Date.now();
const py = await loadPyodide({ stdout: () => {} });          // loader chatter off stdout
await py.loadPackage("micropip", { messageCallback: () => {} });
await py.pyimport("micropip").install(spec);
const t1 = Date.now();
py.setStdout({ batched: (s) => console.log(s) });
py.runPython(readFileSync(join(here, "peer_check.py"), "utf8"));
console.error(`pyodide ${py.version}: ${spec} installed from PyPI in ${((t1 - t0) / 1000).toFixed(1)}s`);

File: demo/pyodide/check.py
#!/usr/bin/env python3
"""Is a browser a peer? Build the same knowledge natively and under Pyodide
(node + the pyodide package, installing ontodag from PyPI) and compare the
canonical roots. Exit 0 iff they agree.

    cd demo/pyodide && npm install && cd -
    python3 demo/pyodide/check.py [VERSION]

VERSION pins the PyPI release Pyodide installs (default: latest). The native
side runs whatever `import ontodag` resolves to, so pass the version that
is installed here when comparing a release rather than a checkout.
"""
import json
import os
import subprocess
import sys

HERE = os.path.dirname(os.path.abspath(__file__))


def main():
    native = json.loads(subprocess.check_output(
        [sys.executable, os.path.join(HERE, "peer_check.py")], text=True))
    cmd = ["node", os.path.join(HERE, "check.mjs")] + sys.argv[1:2]
    wasm = json.loads(subprocess.check_output(cmd, text=True, cwd=HERE))
    for side, r in (("native", native), ("pyodide", wasm)):
        print(f"{side:8} ontodag {r['ontodag']} registry {r['registry']} "
              f"answer {r['answer']} below {r['below']}\n         root {r['root']}")
    if native["root"] == wasm["root"]:
        print("ROOTS AGREE: the browser is a peer")
        return 0
    print("ROOTS DIFFER: something about the encoding differs under wasm")
    return 1


if __name__ == "__main__":
    sys.exit(main())
